internal/planner/consumer.go

```diff
@@ -0,0 +1,320 @@
+	for h, v := range msg.Headers {
+		if strVal, ok := v.(string); ok {
+			headers[h] = strVal
+		}
+	}
+	ctx = otel.GetTextMapPropagator().Extract(ctx, propagation.MapCarrier(headers))
+	ctx, span := trace.Start(ctx, "consume webhook")
+	defer span.End()
+
+	job, err := parseGithubWebhookPayload(ctx, msg.Headers, msg.Body)
+	if err != nil {
+		span.RecordError(err)
+		c.metrics.ObserveWebhookError(ctx, platform)
+		c.discardMessage(ctx, &msg)
+		return err
+	}
+
+	if !slices.Contains([]string{"queued", "in_progress", "completed"}, job.action) {
+		logger.InfoContext(ctx, "ignore other action types for GitHub webhook job", "action", job.action)
+		c.metrics.ObserveDiscardedWebhook(ctx, platform)
+		c.discardMessage(ctx, &msg)
+		return nil
+	}
+
+	if !isSelfHosted(&job) {
+		logger.InfoContext(ctx, "ignore non self-hosted job", "repo", job.repo, "labels", strings.Join(job.labels, ","))
+		c.metrics.ObserveDiscardedWebhook(ctx, platform)
+		c.discardMessage(ctx, &msg)
+		return nil
+	}
+
+	err = c.handleMessage(ctx, &job)
+	if err == nil {
+		c.consumedMessage(ctx, &msg)
+		c.metrics.ObserveConsumedGitHubWebhook(ctx, job.payload)
+		return err
+	}
+
+	span.RecordError(err)
+	c.metrics.ObserveWebhookError(ctx, platform)
+	c.requeueMessage(ctx, &msg)
+	return err
+}
+
+func (c *JobConsumer) discardMessage(ctx context.Context, delivery *amqp.Delivery) {
+	err := delivery.Nack(false, false)
+	if err != nil {
+		logger.ErrorContext(ctx, "cannot discard queue message", "error", err)
+		oteltrace.SpanFromContext(ctx).RecordError(err)
+	}
+	c.metrics.ObserveWebhookError(ctx, platform)
+	return
+}
+
+func (c *JobConsumer) consumedMessage(ctx context.Context, delivery *amqp.Delivery) {
+	err := delivery.Ack(false)
+	if err != nil {
+		logger.ErrorContext(ctx, "cannot ack queue message", "error", err)
+		oteltrace.SpanFromContext(ctx).RecordError(err)
+	}
+	c.metrics.ObserveWebhookError(ctx, platform)
+}
+
+func (c *JobConsumer) requeueMessage(ctx context.Context, delivery *amqp.Delivery) {
+	err := delivery.Nack(false, true)
+	if err != nil {
+		logger.ErrorContext(ctx, "cannot requeue message", "error", err)
+		oteltrace.SpanFromContext(ctx).RecordError(err)
+	}
+	c.metrics.ObserveWebhookError(ctx, platform)
+}
+
+// handleMessage processes a single AMQP message with headers and body.
+func (c *JobConsumer) handleMessage(ctx context.Context, job *githubWebhookJob) error {
+	switch job.action {
+	case "queued":
+		return c.insertJobToDB(ctx, job)
+	case "in_progress":
+		return c.updateJobStartedInDB(ctx, job)
+	case "completed":
+		return c.updateJobCompletedInDB(ctx, job)
+	}
+	return nil
+}
+
+// insertJobToDB inserts a new job into the database.
+func (c *JobConsumer) insertJobToDB(ctx context.Context, job *githubWebhookJob) error {
+	if err := c.db.AddJob(ctx, job.job); err != nil {
+		if errors.Is(err, database.ErrExist) {
+			logger.ErrorContext(ctx, "job already exists in database", "job_id", job.id)
+			return nil
```

Comment:
is there a reason why we return nil? the current implementation `return fmt.Errorf("%w: job %d", ErrJobAlreadyExists, job.GetID())`

Reply:
Yes, that's a good question. I thought about this and, in the end, decided not to classify duplicated webhooks as an error. The first reason is that the duplication can only be detected in `insertJobToDB`, but not in `updateJobStartedInDB` or `updateJobStartedInDB`, so it seems unfair to classify duplicated "queued" webhooks as an error but not "in_progress" or "completed" ones. The second reason is that webhook redelivery is probably expected in webhook systems.